tests/system/test_session_api.py

```diff
@@ -0,0 +1,2126 @@
+    AllTypesRowData(pkey=302, bool_array=[True, False, None]),
+    AllTypesRowData(pkey=303, bytes_array=[BYTES_1, BYTES_2, None]),
+    AllTypesRowData(pkey=304, date_array=[SOME_DATE, None]),
+    AllTypesRowData(pkey=305, float_array=[3.1415926, 2.71828, None]),
+    AllTypesRowData(pkey=306, string_array=["One", "Two", None]),
+    AllTypesRowData(pkey=307, timestamp_array=[SOME_TIME, NANO_TIME, None]),
+    AllTypesRowData(pkey=308, numeric_array=[NUMERIC_1, NUMERIC_2, None]),
+)
+EMULATOR_ALL_TYPES_ROWDATA = (
+    # all nulls
+    EmulatorAllTypesRowData(pkey=0),
+    # Non-null values
+    EmulatorAllTypesRowData(pkey=101, int_value=123),
+    EmulatorAllTypesRowData(pkey=102, bool_value=False),
+    EmulatorAllTypesRowData(pkey=103, bytes_value=BYTES_1),
+    EmulatorAllTypesRowData(pkey=104, date_value=SOME_DATE),
+    EmulatorAllTypesRowData(pkey=105, float_value=1.4142136),
+    EmulatorAllTypesRowData(pkey=106, string_value="VALUE"),
+    EmulatorAllTypesRowData(pkey=107, timestamp_value=SOME_TIME),
+    EmulatorAllTypesRowData(pkey=108, timestamp_value=NANO_TIME),
+    # empty array values
+    EmulatorAllTypesRowData(pkey=201, int_array=[]),
+    EmulatorAllTypesRowData(pkey=202, bool_array=[]),
+    EmulatorAllTypesRowData(pkey=203, bytes_array=[]),
+    EmulatorAllTypesRowData(pkey=204, date_array=[]),
+    EmulatorAllTypesRowData(pkey=205, float_array=[]),
+    EmulatorAllTypesRowData(pkey=206, string_array=[]),
+    EmulatorAllTypesRowData(pkey=207, timestamp_array=[]),
+    # non-empty array values, including nulls
+    EmulatorAllTypesRowData(pkey=301, int_array=[123, 456, None]),
+    EmulatorAllTypesRowData(pkey=302, bool_array=[True, False, None]),
+    EmulatorAllTypesRowData(pkey=303, bytes_array=[BYTES_1, BYTES_2, None]),
+    EmulatorAllTypesRowData(pkey=304, date_array=[SOME_DATE, None]),
+    EmulatorAllTypesRowData(pkey=305, float_array=[3.1415926, 2.71828, None]),
+    EmulatorAllTypesRowData(pkey=306, string_array=["One", "Two", None]),
+    EmulatorAllTypesRowData(pkey=307, timestamp_array=[SOME_TIME, NANO_TIME, None]),
+)
+
+if _helpers.USE_EMULATOR:
+    ALL_TYPES_COLUMNS = EMULATOR_ALL_TYPES_COLUMNS
+    ALL_TYPES_ROWDATA = EMULATOR_ALL_TYPES_ROWDATA
+else:
+    ALL_TYPES_COLUMNS = LIVE_ALL_TYPES_COLUMNS
+    ALL_TYPES_ROWDATA = LIVE_ALL_TYPES_ROWDATA
+
+
+@pytest.fixture(scope="session")
+def sessions_database(shared_instance, operation_timeout):
+    database_name = _helpers.unique_id("test_sessions", separator="_")
+    pool = spanner_v1.BurstyPool(labels={"testcase": "session_api"})
+    sessions_database = shared_instance.database(
+        database_name, ddl_statements=_helpers.DDL_STATEMENTS, pool=pool,
+    )
+    operation = sessions_database.create()
+    operation.result(operation_timeout)  # raises on failure / timeout.
+
+    _helpers.retry_has_all_dll(sessions_database.reload)()
+    # Some tests expect there to be a session present in the pool.
+    pool.put(pool.get())
```

Comment:
I'm not convinced this is needed.

The first test to need a session will trigger session creation and the subsequent tests will reuse the same session. So I don't see a reason to trigger session creation here instead. Is there a session pool related assertion I'm forgetting?

Reply:
Without this line, whether a session has yet been populated depends on the order of invocation of the tests.

The `test_batch_insert_then_read` test [asserts](https://github.com/googleapis/python-spanner/pull/474/files#diff-87e7c347e32cc49ffce2accee7884a79915748a29c3681f1959cb2adfc1e12f3R310-R315) that the first RPC called in the testcase is `GetSession`, not `CreateSession`, and therefore fails if we don't create a session before the OT trace begins for that method.  I guess we could move this line to that test, and then clear the `ot_exporter` afterward to ensure that the `CreateSession` call is not recorded.

Reply:
Another option would be to add a `with_session` fixture, providing this guarantee for `test_batch_insert_then_read` and any other / future tests which expect a session to exist in the pool.

Reply:
Ahhh, I hadn't considered the OT assertions without the guaranteed test order. In that case, I'm okay with this as is :+1:  Thank you for the clear explanation :)